pkg/internal/source/kind.go

```diff
@@ -104,14 +131,61 @@ func (ks *Kind) String() string {
 // WaitForSync implements SyncingSource to allow controllers to wait with starting
 // workers until the cache is synced.
 func (ks *Kind) WaitForSync(ctx context.Context) error {
+	if func() bool {
+		ks.mu.Lock()
+		defer ks.mu.Unlock()
+
+		return ks.started == nil
+	}() {
+		return fmt.Errorf("cannot wait for sync on an unstarted source")
+	}
+
 	select {
 	case err := <-ks.started:
+		if err != nil {
+			// Try to stop the source to clean up any resources
+			// and collect any errors.
+			if stopErr := ks.Stop(); stopErr != nil {
+				err = kerrors.NewAggregate([]error{err, fmt.Errorf("failed to stop source: %w", stopErr)})
+			}
+		}
 		return err
 	case <-ctx.Done():
-		ks.startCancel()
 		if errors.Is(ctx.Err(), context.Canceled) {
 			return nil
 		}
 		return errors.New("timed out waiting for cache to be synced")
 	}
 }
+
+// Stop is internal and should be called only by the Controller to stop the Source. It should block until the
+// Source has stopped.
+func (ks *Kind) Stop() error {
+	// First check that the started channel is populated
+	// by holding the lock.
+	if func() bool {
+		ks.mu.Lock()
+		defer ks.mu.Unlock()
+
+		ks.stopped = true
+
+		return ks.started == nil
+	}() {
+		return nil
+	}
+
+	// Now that we know the started channel is populated,
+	// wait for started to be closed, and then cancel the
+	// context and unregister the handler.
+	var errs []error
+	if err := <-ks.started; err != nil {
+		errs = append(errs, err)
+	}
+	ks.cancel()
+	if ks.removeEventHandler != nil {
```

Comment:
It seems to me that this is not safe indeed. I liked it better when I only had to solve this once because Kind and Informer were sharing the same code.

Reply:
I'll just fix this bug here for now.